File: sdk/runanywhere-swift/Sources/ONNXRuntime/ONNXTTSService.swift
//
//  ONNXTTSService.swift
//  ONNXRuntime
//
//  TTS service implementation using runanywhere-commons ONNX backend.
//  This is a thin Swift wrapper around the rac_tts_onnx_* C APIs.
//

import CRABackendONNX
import CRACommons
import Foundation
import RunAnywhere

/// ONNX-based TTS service for text-to-speech synthesis.
///
/// This service wraps the runanywhere-commons C++ ONNX backend,
/// providing Swift-friendly APIs for speech synthesis using models like Piper.
///
/// ## Usage
///
/// ```swift
/// let service = ONNXTTSService(modelPath: "/path/to/piper-model")
/// try await service.initialize()
///
/// let audioData = try await service.synthesize(
///     text: "Hello, world!",
///     options: TTSOptions(rate: 1.0)
/// )
/// ```
public final class ONNXTTSService: TTSService, @unchecked Sendable {

    // MARK: - Properties

    /// Native handle to the C++ ONNX TTS service
    private var handle: rac_handle_t?

    /// Lock for thread-safe access to handle
    private let lock = NSLock()

    /// Logger for this service
    private let logger = SDKLogger(category: "ONNXTTSService")

    /// Model path for this service
    private let modelPath: String

    /// Whether synthesis is currently in progress
    private var _isSynthesizing = false

    // MARK: - TTSService Protocol

    /// The inference framework (always ONNX)
    public var inferenceFramework: InferenceFramework { .onnx }

    /// Whether synthesis is currently in progress
    public var isSynthesizing: Bool {
        lock.lock()
        defer { lock.unlock() }
        return _isSynthesizing
    }

    /// Available voices (model-dependent)
    public var availableVoices: [String] {
        // For ONNX/Piper models, the model itself is the "voice"
        return [modelPath]
    }

    // MARK: - Initialization

    public init(modelPath: String) {
        self.modelPath = modelPath
        logger.debug("ONNXTTSService instance created for model: \(modelPath)")
    }

    deinit {
        lock.lock()
        if let handle = handle {
            rac_tts_onnx_destroy(handle)
        }
        lock.unlock()
    }

    /// Initialize the service and load the model.
    public func initialize() async throws {
        logger.info("Initializing ONNXTTSService with model: \(modelPath)")

        try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Void, Error>) in
            lock.lock()

            // Clean up existing handle if any
            if let existingHandle = handle {
                rac_tts_onnx_destroy(existingHandle)
                handle = nil
            }

            var newHandle: rac_handle_t?

            // Create config with default values
            var config = rac_tts_onnx_config_t()
            config.num_threads = 4

            // Create service with model
            let result = modelPath.withCString { pathPtr in
                rac_tts_onnx_create(pathPtr, &config, &newHandle)
            }

            if result == RAC_SUCCESS {
                handle = newHandle
                lock.unlock()
                logger.info("ONNXTTSService initialized successfully")
                continuation.resume()
            } else {
                lock.unlock()
                let error = CommonsErrorMapping.toSDKError(result) ?? SDKError.tts(.initializationFailed, "Failed to initialize ONNX TTS service")
                logger.error("Failed to initialize ONNXTTSService: \(error)")
                continuation.resume(throwing: error)
            }
        }
    }

    // MARK: - Synthesis

    /// Synthesize text to audio.
    ///
    /// - Parameters:
    ///   - text: The text to synthesize
    ///   - options: Synthesis options
    /// - Returns: Audio data (16-bit PCM)
    public func synthesize(text: String, options: TTSOptions) async throws -> Data {
        logger.debug("Synthesize called for text: \(text.prefix(50))...")

        return try await withCheckedThrowingContinuation { continuation in
            lock.lock()

            guard let handle = handle else {
                lock.unlock()
                continuation.resume(throwing: SDKError.tts(.serviceNotAvailable, "Service not initialized"))
                return
            }

            _isSynthesizing = true
            lock.unlock()

            defer {
                lock.lock()
                _isSynthesizing = false
                lock.unlock()
            }

            // Convert Swift options to C options
            var cOptions = rac_tts_options_t()
            cOptions.sample_rate = Int32(options.sampleRate)
            cOptions.rate = options.rate

            var cResult = rac_tts_result_t()

            let synthesizeResult = text.withCString { textPtr in
                rac_tts_onnx_synthesize(handle, textPtr, &cOptions, &cResult)
            }

            if synthesizeResult == RAC_SUCCESS {
                // Convert audio data to Data
                var audioData = Data()
                if let dataPtr = cResult.audio_data, cResult.audio_size > 0 {
                    // Audio data is already in the correct format
                    audioData = Data(bytes: dataPtr, count: cResult.audio_size)
                    // Free the C-allocated data
                    dataPtr.deallocate()
                }

                logger.debug("Synthesis complete: \(cResult.audio_size) bytes")
                continuation.resume(returning: audioData)
            } else {
                let error = CommonsErrorMapping.toSDKError(synthesizeResult) ?? SDKError.tts(.generationFailed, "Synthesis failed")
                logger.error("Synthesis failed: \(error)")
                continuation.resume(throwing: error)
            }
        }
    }

    /// Stream synthesis for long text.
    ///
    /// Note: ONNX TTS backend does not support streaming synthesis.
    /// This method synthesizes the full text and returns it as a single chunk.
    ///
    /// - Parameters:
    ///   - text: The text to synthesize
    ///   - options: Synthesis options
    ///   - onChunk: Callback for each audio chunk
    public func synthesizeStream(
        text: String,
        options: TTSOptions,
        onChunk: @escaping (Data) -> Void
    ) async throws {
        logger.debug("Stream synthesize called for text: \(text.prefix(50))...")

        // ONNX TTS doesn't support streaming - synthesize full audio and return as single chunk
        let audioData = try await synthesize(text: text, options: options)
        onChunk(audioData)

        logger.debug("Stream synthesis complete (non-streaming fallback)")
    }

    // MARK: - Lifecycle

    /// Stop current synthesis
    public func stop() {
        lock.lock()
        if let handle = handle {
            rac_tts_onnx_stop(handle)
            logger.debug("Synthesis stopped")
        }
        _isSynthesizing = false
        lock.unlock()
    }

    /// Clean up resources
    public func cleanup() async {
        // Use synchronous cleanup to avoid NSLock in async context issues
        cleanupSync()
    }

    /// Synchronous cleanup helper
    private func cleanupSync() {
        lock.lock()
        if let handle = handle {
            rac_tts_onnx_destroy(handle)
            self.handle = nil
            logger.info("ONNXTTSService cleaned up")
        }
        lock.unlock()
    }
}

File: sdk/runanywhere-flutter/packages/runanywhere_onnx/lib/services/onnx_tts_service.dart
import 'dart:async';
import 'dart:typed_data';

import 'package:runanywhere/core/module_registry.dart';
import 'package:runanywhere/native/native_backend.dart';

/// ONNX-based Text-to-Speech service.
///
/// This is the Flutter equivalent of Swift's `ONNXTTSService`.
/// It uses the native runanywhere-core library for speech synthesis.
///
/// ## Usage
///
/// ```dart
/// final backend = NativeBackend.onnx();
/// final tts = OnnxTTSService(backend);
/// await tts.initialize(modelPath: '/path/to/model');
///
/// final input = TTSInput(text: 'Hello, world!');
/// final output = await tts.synthesize(input);
/// ```
class OnnxTTSService implements TTSService {
  final NativeBackend _backend;
  bool _isInitialized = false;
  bool _isSynthesizing = false;
  List<String> _voices = [];

  /// Create a new ONNX TTS service.
  OnnxTTSService(this._backend);

  /// Get the inference framework
  String get inferenceFramework => 'onnx';

  @override
  bool get isReady => _isInitialized && _backend.isTtsModelLoaded;

  /// Whether synthesis is currently in progress
  bool get isSynthesizing => _isSynthesizing;

  /// Get available voice IDs
  List<String> get availableVoices => _voices;

  @override
  Future<void> initialize({String? modelPath}) async {
    if (modelPath == null || modelPath.isEmpty) {
      _isInitialized = true;
      return;
    }

    try {
      // Load the TTS model through native backend
      _backend.loadTtsModel(
        modelPath,
        modelType: 'vits',
      );

      // Verify the model loaded successfully
      if (!_backend.isTtsModelLoaded) {
        throw Exception('TTS model failed to load');
      }

      // Get available voices
      _voices = _backend.getTtsVoices();
    } catch (e) {
      _isInitialized = false;
      rethrow;
    }

    _isInitialized = true;
  }

  @override
  Future<TTSOutput> synthesize(TTSInput input) async {
    if (!isReady) {
      throw Exception('TTS service not ready');
    }

    _isSynthesizing = true;

    try {
      final result = _backend.synthesize(
        input.text,
        voiceId: input.voiceId,
        speed: input.rate,
        pitch: input.pitch - 1.0, // Convert from 0.5-2.0 range to -0.5-1.0
      );

      final samples = result['samples'] as Float32List;
      final sampleRate = result['sampleRate'] as int;

      // Convert Float32 samples to PCM16 bytes
      final audioData = _convertToPCM16(samples);

      return TTSOutput(
        audioData: audioData,
        format: 'pcm',
        sampleRate: sampleRate,
      );
    } finally {
      _isSynthesizing = false;
    }
  }

  @override
  Future<void> cleanup() async {
    if (_backend.isTtsModelLoaded) {
      _backend.unloadTtsModel();
    }
    _isInitialized = false;
    _voices = [];
  }

  // ============================================================================
  // Private Helpers
  // ============================================================================

  List<int> _convertToPCM16(Float32List samples) {
    final pcm16 = Uint8List(samples.length * 2);

    for (var i = 0; i < samples.length; i++) {
      // Clamp to -1.0 to 1.0
      final clamped = samples[i].clamp(-1.0, 1.0);
      // Convert to 16-bit signed integer
      final sample = (clamped * 32767).round();
      // Store as little-endian
      pcm16[i * 2] = sample & 0xFF;
      pcm16[i * 2 + 1] = (sample >> 8) & 0xFF;
    }

    return pcm16;
  }
}
